src/Processors/Formats/Impl/FreeformRowInputFormat.cpp

```diff
@@ -0,0 +1,473 @@
+            columns, matchers.size() - 1, scoreForField(matchers.back()->getEscapingRule(), columns[0].type), in.position(), false);
+        return next_fields;
+    }
+
+    size_t best_type_score = 0, type_score = 0, fields_score = 0;
+    for (size_t i = 0; const auto & matcher : matchers)
+    {
+        try
+        {
+            auto fields = matcher->parseFields(in, index);
+            NamesAndTypes columns; // we either add all of the fields or none of them
+            columns.reserve(fields.size());
+            type_score = 0;
+            fields_score = 0;
+
+            for (const auto & [name, field] : fields)
+            {
+                auto type = matcher->getDataTypeFromField(field);
+                if (type)
+                {
+                    // a single punctuation is not a valid field
+                    if (field.size() == 1 && isPunctuationASCII(field[0]))
+                        continue;
+
+                    columns.emplace_back(name, type);
+
+                    type_score += scoreForType(type);
+                    fields_score += scoreForField(matcher->getEscapingRule(), type);
+
+                    one_string = field.ends_with(':') && matcher->getName() == "RawByWhitespaceFieldMatcher";
+                }
+            }
+
+            // add a new field only if type_score > best_type_score or if we haven't found a field other than string
+            if (columns.size() == fields.size() && (type_score > best_type_score || best_type_score <= 1))
+            {
+                best_type_score = type_score;
+                next_fields.emplace_back(columns, i, fields_score, in.position(), one_string);
+            }
+        }
+        catch (Exception & e)
+        {
+            LOG_DEBUG(&Poco::Logger::get("FreeformFieldMatcher"), "fail to parse field: {}", e.message());
+        }
+
+        ++i;
+        in.position() = start;
+    }
+
+    return next_fields;
+}
+
+void FreeformFieldMatcher::recursivelyGetNextFieldInRow(
+    char * current_pos, Solution current_solution, std::vector<Solution> & solutions, bool one_string)
+{
+    char * tmp = in.position();
+    in.position() = current_pos;
+    if (*in.position() == '\n')
+    {
+        solutions.push_back(current_solution);
+        return;
+    }
+
+    const auto next_fields = readNextFields(one_string, current_solution.size);
+    for (const auto & fields : next_fields)
+    {
+        auto next = current_solution;
+        next.matchers_order.push_back(fields.matcher_index);
+        for (const auto & column : fields.columns)
+            next.columns.push_back(column);
+
+        next.score += fields.score;
+        next.size += fields.columns.size();
+
+        recursivelyGetNextFieldInRow(fields.pos, next, solutions, fields.parse_the_rest_as_one_string);
+    }
+
+    in.position() = tmp; // reset to initial position
+}
+
+void FreeformFieldMatcher::readRowAndGenerateSolutions(char * pos, std::vector<Solution> & solutions)
```

Comment:
It was a bit difficult to understand all details and logic. Please, correct me if I missed something.
What I understood so far (please correct me):
We remember current position of the buffer and iterate over our matchers. For each matcher we set position to the current start and try to parse a field (or fields if we have JSON object). Then we choose the best matched fields and remember the end of these fields in the buffer. Then we iterate over these best fields, update current solution and call it recursively.

It looks ok, but we will have problems with buffer as I explained in one of comments above. Also, it would be good to refactor code a bit and add more comments to make it more readable. I expect it to be smth like this (it doesn't mean to write exactly like this, I just want to show an examle):
```c++
std::vector<Fields> readNextFields(PeekableReadBuffer & in)
{
    in.setCheckpoint();
    size_t best_score = 0;
    std::vector<Fields> result_fields;
    for (auto & matcher : matchers)
    {
        auto result = matcher->readFields(in);   /// result can contain information about types, scores and some other information about fields (like it’s size, is it a valid field etc)
        if (!result.ok)
            continue;
        if (best_score <= 1 || result.score > best_score)
        {
            best_score = result.score;
            /// We cannot save position in the buffer, because buffer can be changed.
            /// But we can remember how much data were read since last checkpoint.
            /// To do it, we can add new method to PeekableReadBuffer like in.offsetFromLastCheckpoint()
            size_t offset = in.offsetFromLastCheckpoint();
            result_fields.emplace_back(matcher, result, offset); /// matcher can be unique_ptr, ok to copy it instead of index.
        }
        in.rollbackToCheckpoint();
    }
    in.dropCheckpoint();
}

void buildSolutions(PeekableReadBuffer & in, Solution & current_solution, Solutions & solutions)
{
    if (in.eof() || *in.position() == '\n')
        solutions.push_back(current_solution);

    auto next_fields = readNextFields(in);

    in.setCheckpoint();
    for (const auto & next_field : next_fields)
    {
        auto new_solution = current_solution;
        new_solution.addMatcher(next_field.matcher);
        new_solution.addFields(nest_field.result.fields);
        new_solution.score += nest_field.result.score;
        /// Set buffer position to the end of current fields.
        in.ignore(next_fields.offset);
        buildSolutions(in, new_solution, solutions);
        in.rollbackToCheckpoint();
    }
    in.dropCheckpoint();
}
```

Reply:
Yes, that looks right, I will refactor the code a bit more and test it